### ISSUE TITLE
Inconsistent canSee behaviour between showOnTableRow and on detail

### ISSUE THREAD

**Issue by rdarcy1** [laravel/nova-issues#2655]
- Laravel Version: 7.13.0
- Nova Version: 3.6.0
- PHP Version: 7.4.5
- Database Driver & Version: MySQL 5.7
- Operating System and Version: MacOS Catalina
- Browser type and version: N/A

### Description:
When authorising an action per-resource via `canSee`, the model must be accessed in a different way when on table row vs on its detail page. 

This may be related to #1943 and #2221, but I think it's a distinct issue because the behaviour is there for both cases – it's just different (which requires some nasty context checking).

### Steps To Reproduce:
- Given a `Payment` model, with a `paid_at` datetime, and an action `MarkPaymentPaid` that populates `paid_at` with the current date.
- The action should only be visible if the `paid_at` field is `null`.
- Assume that the payment's `paid_at` field has already been set, so the action shouldn't show.

#### Works correctly on detail
The following code (written per the docs) works on detail, but will incorrectly show the action on the table row, because `$request->findModelQuery()->first()` returns `null` on table row.

```php
public function actions(Request $request)
{
    return [
        (new MarkPaymentPaid)->showOnTableRow()
            ->canSee(function (NovaRequest $request) {
                return ! optional($request->findModelQuery()->first())->paid_at;
            }),
    ];
}
```

#### Works correctly on table row
This works on the table row, but will incorrectly show the action on detail, because `$this->resource` returns an empty `Payment` model on detail.

```php
public function actions(Request $request)
{
    return [
        (new MarkPaymentPaid)->showOnTableRow()
            ->canSee(function (NovaRequest $request) {
                return ! $this->resource->paid_at;
            }),
    ];
}
```

#### Works correctly on both
A workaround is to check for a blank `$this->resource` and use the query instead.
```php
public function actions(Request $request)
{
    return [
        (new MarkPaymentPaid)->showOnTableRow()
            ->canSee(function (NovaRequest $request) {
                return ! optional($this->resource->exists ? $this->resource : $request->findModelQuery()->first())->paid_at;
            }),
    ];
}
```

This seems to work OK for now, but doesn't seem like a good solution because:
- The closure shouldn't care about the context in which it's called
- It's (unnecessarily) verbose
- It's not documented, and I assume it could change in the future if Nova changes how it deals with the request internally

**Comment by rdarcy1** [comment]
Still an issue in Nova 3.8.2

**Comment by rdarcy1** [comment]
Bump

**Comment by jonnott** [comment]
Bump

**Comment by worthers** [comment]
Bump. We have the same problem - canSee cannot be used at all in the context of the index view since there is no way to query 'current' item.

**Comment by hnphi** [comment]
In my `App\Nova\User.php` 

```
 public function actions(Request $request)
    {
        $active = optional($this->resource->exists ? $this->resource : $request->findModelQuery()->first())->active;

        return [
            (new Deactive)
                ->showOnTableRow()
                ->canSee(function (NovaRequest $request) use($active) {
                    return $active;
                })
                ->withoutConfirmation(),
            (new Active)
                ->showOnTableRow()
                ->canSee(function (NovaRequest $request) use($active) {
                    return !$active;
                })
                ->withoutConfirmation(),
        ];
    }
```

Action `Active` set User's attribute `active` = `true`, otherwise set attribute `active` = `false` with `Deactive` action, so I make logic: 

- if User's `active` is `true` => only show button `Deactive`
- if User's `active` is `false` => only show button `Active`

action button `Active` work fine, but `Deactive` get `403 error`.

I was check file `nova\src\Http\Requests\ActionRequest.php`: function `availableActions()` return empty array if action `Deactive` to be call.

**Comment by jonnott** [comment]
@davidhemphill Does the marking as bug and closing of this issue mean it's been fixed in v3.14 ?

**Comment by crynobone** [comment]
Hi @jonnott 

> Accessing `/nova-api/users/actions?resourceId=2` should attempt to create resource with User model `id=2` and not `null` model. 

This fixed the issue where `index` and `detail` require different logic to gain access to the `resource`.

### Previous behaviour

```php
public function actions(Request $request)
{
    return [
        (new MarkPaymentPaid)->showOnTableRow()
            ->canSee(function (NovaRequest $request) {
                return ! optional($this->resource->exists ? $this->resource : $request->findModelQuery()->first())->paid_at;
            }),
    ];
}
```

### New behaviour

```php
public function actions(Request $request)
{
    return [
        (new MarkPaymentPaid)->showOnTableRow()
            ->canSee(function (NovaRequest $request) {
                return $request instanceof ActionRequest 
                    || ($this->resource->exists && $this->resource->paid_at); 
            }),
    ];
}
```

### Limitation

`$request instanceof ActionRequest` is still required as of `v3.14` to allow running the action. Ideally I would like to remove `canSee()` requirement when executing the action since it should just use `canRun()`. Will come out with a proposal on that soon.
